### PR TITLE
[JIT Datasources] Restrict data source block access to conversations space II

### PR DESCRIPTION
**Pull request by philipperolet**
Description
---
Fixes https://github.com/dust-tt/tasks/issues/1658 (again)

Following #8815 , @Fraggle  identified an issue described [here](https://dust4ai.slack.com/archives/C080VUG3PQR/p1732275137389959?thread_ts=1732271409.175599&cid=C080VUG3PQR)

A new approach is taken: pass a flag to core indicating if the call is system. Core can then ask the registry to allow or not datasources from the "conversations" space.

- for the front v1 app/runs -> core call, adding the flag as a dust header. It's immutably decided by the nature of the API key (and feels more secure than leaving it open to callers)
- for the lookup call, adding as a param to the API query, it seems like a parameter callers from core might want to set / unset

Risk
---
none, not yet used in prod (right @Fraggle ?)

Deploy
---
front
core

### CODE REVIEW

**Comment by Fraggle** [comment]
>none, not yet used in prod (right @Fraggle ?)

Right :)

**Review by Fraggle — APPROVED** [review]
LGTM.

I'm a bit bugged by:
`isSystemKey` => `X-Dust-IsSystem` => `DUST_IS_SYSTEM_RUN` => `allow_conversations_data_sources`

And that core knows about "conversations" data sources.

**Comment by philipperolet** [comment]
> LGTM.
> 
> I'm a bit bugged by: `isSystemKey` => `X-Dust-IsSystem` => `DUST_IS_SYSTEM_RUN` => `allow_conversations_data_sources`
> 
> And that core knows about "conversations" data sources.

Yep I understand, it looks weird this way. I would explain like this:
1. I would agree to rename `X-Dust-IsSystem` to `X-Dust-IsSystemRun`
2. Then, the API endpoint is a "run" endpoint => it sends headers about runs. It determined it was a system run by the fact that a system key was used => this transition thus seems natural
3. lookup should not know about callers details, just whether it should return the datasource or a not found => thus the allow_conversations_datasource

I agree that it's surprising to have the word conversations in core, but it's actually not really "conversations" per se, but the space conversations that's designated; so we had already somewhat crossed that border when we added DUST_GROUP_IDS in the env

We could discuss rename to "allow_system_data_source" or something, maybe rename everything "allow_restricted_spaces" ("conversations" would be one of them)

File: core/src/blocks/data_source.rs [hunk 1]
```diff
@@ -82,11 +82,21 @@ impl DataSource {
         filter: Option<SearchFilter>,
         target_document_tokens: Option<usize>,
     ) -> Result<Vec<Document>> {
+        // This check is meant to block access to "conversations" space through a
+        // datasource block in a dust app, which could lead to data leaks, see related issue
+        // https://github.com/dust-tt/tasks/issues/1658
+        // Only case in which this is allowed is for our packaged apps, via a system
+        // key, in particular "assistant-retrieval-v2" that needs access to the
+        // conversation space
```
Comment (spolu) [inline core/src/blocks/data_source.rs:90]: Can we just convey isSystem and not say anyting about conversations data source here :pray:

File: core/src/blocks/helpers.rs [hunk 2]
```diff
@@ -47,11 +50,13 @@ pub async fn get_data_source_project_and_view_filter(
     };
 
     let url = format!(
-        "{}/api/registry/data_sources/lookup?workspace_id={}&data_source_id={}",
+        "{}/api/registry/data_sources/lookup?workspace_id={}&data_source_id={}&allow_conversations_datasources={}",
```
Comment (spolu) [inline core/src/blocks/helpers.rs:53]: allow_conversations_datasources => is_system_run

**Pull request merged**